### PR TITLE
Fixing standalone mode

### PR DESCRIPTION
**Pull request by zapu**
Here is what I have figured out so far. Seems to work. I'm not sure if `StartStandaloneChat` contains bare minimum to connect to gregor and setup chat subsystem, or if it's missing anything. I'll be happy to refactor if this code should be structured differently. Thanks!

This is an ongoing effort, there are still more commands that fail and will be either fixed or `SetNoStandalone`'d.

### CODE REVIEW

File: go/libkb/globals.go [hunk 1]
```diff
@@ -1017,3 +1023,16 @@ func (g *GlobalContext) ClearPerUserKeyring() {
 func (g *GlobalContext) LocalSigchainGuard() *LocalSigchainGuard {
 	return g.localSigchainGuard
 }
+
+func (g *GlobalContext) StartStandaloneChat() {
+	if g.Service {
```
Comment (zapu) [inline go/libkb/globals.go]: this turns out to be not a good way to test for (not) standalone... any hints?

**Comment by zapu** [comment]
This is not great, though, `chat send` for example will always try to create "new conversation" - it's possible that it's unable to Resolve existing one.

File: go/service/teams.go [hunk 2]
```diff
@@ -139,6 +139,10 @@ func (h *TeamsHandler) TeamAddMember(ctx context.Context, arg keybase1.TeamAddMe
 		return result, nil
 	}
 
+	// Ensure we have chat available, since TeamAddMember may also be
+	// coming from a standalone launch.
+	h.G().ExternalG().StartStandaloneChat()
```
Comment (mmaxim) [inline go/service/teams.go]: Can this go in `sendTeamChatWelcomeMessage`?

File: go/libkb/globals.go [hunk 3]
```diff
@@ -109,6 +114,8 @@ type GlobalContext struct {
 	ConnectivityMonitor ConnectivityMonitor  // Detect whether we're connected or not.
 	localSigchainGuard  *LocalSigchainGuard  // Non-strict guard for shoeing away bg tasks when the user is doing sigchain actions
 
+	SChatConnector StandaloneChatConnector
```
Comment (mmaxim) [inline go/libkb/globals.go]: This is kind of a weird name, why not just name `StandaloneChatConnector`?

File: go/client/cmd_chat_send.go [hunk 4]
```diff
@@ -76,6 +76,16 @@ func (c *CmdChatSend) Run() (err error) {
 	if c.resolvingRequest.Visibility == chat1.TLFVisibility_ANY {
 		c.resolvingRequest.Visibility = chat1.TLFVisibility_PRIVATE
 	}
+
+	if c.G().Standalone {
+		if c.resolvingRequest.MembersType == chat1.ConversationMembersType_KBFS {
```
Comment (mmaxim) [inline go/client/cmd_chat_send.go]: make this `switch`, since we'll have `IMPTEAM` in here at some point

File: go/client/cmd_chat_send.go [hunk 5]
```diff
@@ -76,6 +76,16 @@ func (c *CmdChatSend) Run() (err error) {
...
+			err = fmt.Errorf("Can't run command in standalone mode")
```
Comment (mmaxim) [inline go/client/cmd_chat_send.go]: Is there a generic error we could return here for no standalone mode?

File: go/service/teams.go (hunk 2 above)
Comment (zapu) [inline go/service/teams.go]: the global context is not passed through there, should I pass it?

File: go/client/cmd_chat_send.go (hunk 5 above)
Comment (maxtaco) [inline go/client/cmd_chat_send.go]: I agree, we should have a generic error for this condition.

File: go/client/cmd_chat_send.go [hunk 6]
```diff
@@ -76,6 +76,17 @@ func (c *CmdChatSend) Run() (err error) {
 	if c.resolvingRequest.Visibility == chat1.TLFVisibility_ANY {
 		c.resolvingRequest.Visibility = chat1.TLFVisibility_PRIVATE
 	}
+
+	if c.G().Standalone {
+		switch c.resolvingRequest.MembersType {
+		case chat1.ConversationMembersType_TEAM:
+			c.G().StartStandaloneChat()
+		default:
+			err = fmt.Errorf("Can't run command in standalone mode")
```
Comment (maxtaco) [inline go/client/cmd_chat_send.go]: use a special error for this

Comment (zapu) [inline go/client/cmd_chat_send.go]: working on it, thank you all for looking!

File: go/service/teams.go (hunk 2 above)
Comment (mmaxim) [inline go/service/teams.go]: It should be on h.G() right?

Comment (zapu) [inline go/service/teams.go]: You're right, sorry

File: go/client/errors.go [hunk 7]
```diff
@@ -70,3 +70,9 @@ type UnexpectedArgsError string
 func (e UnexpectedArgsError) Error() string {
 	return fmt.Sprintf("Command `%s` doesn't take any non-flag arguments", string(e))
 }
+
+type CantRunInStandaloneError struct{}
+
+func (e CantRunInStandaloneError) Error() string {
+	return fmt.Sprintf("Can't run command in standalone mode")
```
Comment (maxtaco) [inline go/client/errors.go]: just return the string without `sprintf`?

**Comment by zapu** [comment]
ok, thank you

**Pull request merged**